typescript-rewrite/samlang-core/src/interpreter/expression-interpreter.ts

```diff
@@ -0,0 +1,274 @@
+import { IdentifierType } from '../ast/common/types';
+import { SamlangExpression } from '../ast/lang/samlang-expressions';
+import { InterpretationContext, EMPTY } from './interpretation-context';
+import PanicException from './panic-exception';
+import {
+  Value,
+  ObjectValue,
+  IntValue,
+  BoolValue,
+  StringValue,
+  FunctionValue,
+  VariantValue,
+  TupleValue,
+  isSameValue,
+} from './value';
+
+export default class ExpressionInterpreter {
+  private printedCollector = '';
+
+  private blameTypeChecker = (message = ''): never => {
+    throw Error(message);
+  };
+
+  readonly eval = (
+    expression: SamlangExpression,
+    context: InterpretationContext = EMPTY
+  ): Value => {
+    switch (expression.__type__) {
+      case 'LiteralExpression': {
+        switch (expression.literal.type) {
+          case 'IntLiteral':
+            return { type: 'int', value: Number(expression.literal.value) };
+          case 'StringLiteral':
+            return { type: 'string', value: expression.literal.value };
+          case 'BoolLiteral':
+            return { type: 'bool', value: expression.literal.value };
+        }
+      }
+      // eslint-disable-next-line no-fallthrough
+      case 'ThisExpression':
+        return context.localValues.get('this') ?? this.blameTypeChecker('Missing `this`');
+      case 'VariableExpression':
+        return (
+          context.localValues.get(expression.name) ??
+          this.blameTypeChecker(`Missing variable ${expression.name}`)
+        );
+      case 'ClassMemberExpression':
+        return (
+          context.classes.get(expression.className)?.functions?.get(expression.memberName) ??
+          this.blameTypeChecker()
+        );
+      case 'TupleConstructorExpression':
+        return {
+          type: 'tuple',
+          tupleContent: expression.expressions.map((e) => this.eval(e, context)),
+        };
+      case 'ObjectConstructorExpression': {
+        const objectContent = new Map<string, Value>();
+        expression.fieldDeclarations.forEach((declaration) => {
+          if (declaration.expression) {
+            objectContent.set(declaration.name, this.eval(declaration.expression, context));
+          }
```

Comment:
This is wrong. When expression is undefined, it refers to this case:

```js
val a = { foo, bar };
```

The semantics is same as JS when you can omit the expression if the expression is the a variable with the same name as the field name

Reply:
Added the else branch to set a variable